### PR TITLE
Fixes #26073 - Adding react infrastructure

### PR DESCRIPTION
**Pull request by Laviro**

### CODE REVIEW

**Comment by sharvit** [comment]
**TODO:**
- [ ] Use babel configurations as core uses
- [ ] Use eslint configurations as core uses
- [ ] Use travis configurations as core uses
- [ ] Use Jest configurations as core uses
- [ ] Create hello-world react-component
- [ ] Test the hello-world component with Jest
- [ ] Add storybook with the hello-world component

File: webpack/global_test_setup.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// runs before each test to make sure console.error output will
+// fail a test (i.e. default PropType missing). Check the error
+// output and traceback for actual error.
+global.console.error = (error) => {
```
Comment (sharvit) [inline webpack/global_test_setup.js]: I don't sure we actually need this file, in core it just works fine

File: webpack/.eslintrc [hunk 2]
```diff
@@ -0,0 +1,35 @@
+{
```
Comment (sharvit) [inline webpack/.eslintrc]: Better to use the core configurations that use pf-react-eslint

File: webpack/global_test_setup.js (hunk 1 above)
Comment (Laviro) [reply webpack/global_test_setup.js]: thanks, removed it and instead using only `test_setup.js`

File: webpack/.eslintrc (hunk 2 above)
Comment (Laviro) [reply webpack/.eslintrc]: using `patternfly-react-eslint` now

**Comment by sharvit** [comment]
Based on this core pr https://github.com/theforeman/foreman/pull/6515

File: app/controllers/foreman_tasks/react_controller.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+module ForemanTasks
+  class ReactController < ::ApplicationController
```
Comment (ohadlevy) [inline app/controllers/foreman_tasks/react_controller.rb]: why is that needed?

File: package.json [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    "stylelint-config-standard": "^18.0.0",
+    "uglifyjs-webpack-plugin": "^1.2.2",
+    "url-loader": "^1.0.1",
+    "webpack": "^3.4.1",
+    "webpack-bundle-analyzer": "^2.13.1",
+    "webpack-dev-server": "^2.5.1",
+    "webpack-stats-plugin": "^0.1.5"
+  },
+  "dependencies": {
+    "@novnc/novnc": "^1.0.0",
```
Comment (ohadlevy) [inline package.json]: please don't copy all of the packages from foreman as they are not all needed here ^

Comment (Laviro) [reply package.json]: thanks, sure.. it's just that many things broke when I removed some of them.. so i wanted to make sure everything works before.

**Comment by Laviro** [comment]
Storybook is working now:
selection_018

**Comment by Laviro** [comment]
if you'll go to the `/foreman_tasks/ex_tasks` (experimental page), you will see:
selection_020

File: app/controllers/foreman_tasks/react_controller.rb (hunk 3 above)
Comment (sharvit) [reply app/controllers/foreman_tasks/react_controller.rb]: We want to implement it with `react-router`

**Comment by Laviro** [comment]
@sharvit thanks for all of your commits in this PR !!

File: app/controllers/foreman_tasks/react_controller.rb (hunk 3 above)
Comment (ohadlevy) [reply app/controllers/foreman_tasks/react_controller.rb]: ack, lets finalize the way its supposed to be used in foreman first? ( i would assume as a plugin you would reuse that infra)

Comment (sharvit) [reply app/controllers/foreman_tasks/react_controller.rb]: agree, imo foreman should provide `registerRoutes` so we won't need to use turbolinks when moving between foreman-routes to plugin-routes.

**Comment by Laviro** [comment]
@sharvit Thanks for your help on this, couldn't do this without you

**Comment by iNecas** [comment]
Awesomesauce. I need to test it + figure out how to skip the auth tests, but hopefully could get in by tomorrow.

File: app/controllers/foreman_tasks/react_controller.rb (hunk 3 above)
Comment (waldenraines) [reply app/controllers/foreman_tasks/react_controller.rb]: @sharvit how much work would it be to do the above?

File: app/controllers/foreman_tasks/react_controller.rb [hunk 5]
```diff
@@ -1,11 +1,17 @@
 module ForemanTasks
   class ReactController < ::ApplicationController
-    skip_before_action :authorize
```
Comment (iNecas) [inline app/controllers/foreman_tasks/react_controller.rb]: Added authorization, to make tests green, but it also probably make sense to not expose it without having tasks permissions (although the permissions will be covered by API)

**Comment by iNecas** [comment]
Travis has been enabled for this repo

**Comment by Laviro** [comment]
ESlint is working now,, 
I saw how Katello managed this issue.

**Comment by iNecas** [comment]
Awesome. I've pushed one more fix regarding permissions + rebased against master because there were some tests failing due to other issue we've fixed in the mean time. Once tests are green, this will go in.

File: lib/foreman_tasks/engine.rb [hunk 6]
```diff
@@ -57,7 +57,9 @@ class Engine < ::Rails::Engine
 
         security_block :foreman_tasks do |_map|
           permission :view_foreman_tasks, { :'foreman_tasks/tasks' => [:auto_complete_search, :sub_tasks, :index, :show],
-                                            :'foreman_tasks/api/tasks' => [:bulk_search, :show, :index, :summary] }, :resource_type => ForemanTasks::Task.name
+                                            :'foreman_tasks/react' => [:index],
+                                            :'foreman_tasks/api/tasks' => [:bulk_search, :show, :index, :summary],
```
Comment (adamruzicka) [inline lib/foreman_tasks/engine.rb]: Rubocop is not happy about changes in this file

Comment (iNecas) [reply lib/foreman_tasks/engine.rb]: Should be fixed now

**Comment by iNecas** [comment]
All almost green (test failure unrelated). Thanks @Laviro @sharvit @ohadlevy

**Comment by iNecas** [comment]
Squashed and merged in a040206
